chore: data model cleanup

ErikApption opened the pull request:
# Pull Request

## Commit Title
- Pre-requisite for major External User data model changes
- Enabled nullable in `DataHub.Core` project for modern C# code
- Major refactoring to validate required fields
- Setup `required` on all required fields
- Populated required fields through solution

## Testing

- Tested locally

## Checklist
<!-- Ensure the following tasks are complete -->

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [X] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [X] Added `!` to the type prefix if change is a breaking change (i.e., requires minor or major version bump)
- [X] Targeted the correct branch (see [External Collaboration](https://github.com/fsdh-pfds/.github/blob/main/CONTRIBUTING.md#external-collaboration))
- [X] ensured that my code follows coding standards
- [X] written tests and theyre passing


### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage

ErikApption committed:
fix for relation bug in projects

ErikApption committed:
removed primary attribute

ErikApption committed:
added migration for new cardinality

ErikApption committed:
added  sql script to update identity in migration

ErikApption committed:
merged develop

ErikApption committed:
repaired migrations

ErikApption committed:
initial unit tests for migrations

ErikApption committed:
refactored all users stuff into users namespace

ErikApption committed:
nullable wip1

ErikApption committed:
fixed core

ErikApption committed:
fioxes

ErikApption committed:
FIXED ALL COMPILATION ISSUES 🥳🥳🥳

ErikApption committed:
removed unused cloud provider field and moved public data fields into nullable

ErikApption committed:
temporarily removed external user for future migration

ErikApption committed:
merged develop

ErikApption committed:
merged develop

ErikApption committed:
merged develop

ErikApption committed:
improve resource group logic for unit tests

ErikApption committed:
- fixed unit tests
- added migration for minor model changes

Sean-Stilwell reviewed (dismissed):
Looks good to me after review & local testing.

akhavich commented on Portal/src/Datahub.Core/Components/AnchorLink.razor:
```diff
@@ -63,4 +62,4 @@
 
-<a @attributes="this.Attributes" @onclick="this.AnchorOnClickAsync" @onclick:preventDefault="@this.preventDefault">@this.ChildContent</a>
\ No newline at end of file
+<a @attributes="this.Attributes" @onclick="this.AnchorOnClickAsync" @onclick:preventDefault="@this.preventDefault">@this.ChildContent</a><a @attributes="this.Attributes" @onclick="this.AnchorOnClickAsync" @onclick:preventDefault="@this.preventDefault">@this.ChildContent</a>
```
Is this meant to be duplicated?

ErikApption replied:
thanks for catching this

gcnabeel commented on Portal/src/Datahub.Core/Components/FileUpload/DropZone.razor:
```diff
@@ -58,8 +58,7 @@
     public void Dispose()
     {
-        ((IDisposable) _inputFile)?.Dispose();
         _dotNetHelper?.Dispose();
```
you may want to dispose off the js objects here:

@implements IAsyncDisposable (for async madness)
 
 if (_dropZoneInstance is not null)
        await _dropZoneInstance.DisposeAsync();

    if (_module is not null)
        await _module.DisposeAsync();

akhavich commented on Portal/src/Datahub.Core/Services/Api/ApiTelemetryService.cs:
```diff
@@ -10,7 +10,7 @@ public class ApiTelemetryService
     private readonly TelemetryConfiguration _config = TelemetryConfiguration.CreateDefault();
 
-    private TelemetryClient _client;
+    private TelemetryClient? _client;
```
Since this gets initialized in the constructor id shouldn't need to be nullable. We can also get rid of the null check in the Client getter, or maybe get rid of _client entirely and just have protected TelemetryClient Client {get; private set;}

ErikApption replied:
another good catch thanks

akhavich commented on Portal/src/Datahub.Core/Services/TranslationService.cs:
```diff
@@ -11,7 +11,7 @@ public TranslationService(IConfiguration configuration)
         var useFreeApi = configuration.GetSection("DeepL").GetValue("UseFreeApi", true);
         var authKey = configuration.GetSection("DeepL").GetValue<string>("AuthKey");
-        _client = new Translator(authKey, new TranslatorOptions() { ServerUrl = useFreeApi ? "https://api-free.deepl.com" : null });
+        _client = new Translator(authKey!, new TranslatorOptions() { ServerUrl = useFreeApi ? "https://api-free.deepl.com" : null });
```
We may not need to specify the server url at all; according to the Deepl-dotnet documentation it will auto select the correct api url based on the user's account. It's not critical but we can add a TODO to revisit it later

akhavich commented on Portal/src/Datahub.Core/Templates/TestEmailTemplate.razor:
I don't think this file is used any more (since nrcan!) but on the off chance it is it should at least be updated to SSC, or deleted entirely

ErikApption committed:
- addressed copilot feedback with repeated import
- addressed disposable
- corrected nullables for databricks API and added references

akhavich commented on Portal/src/Datahub.Portal/Layout/PersonalSidebar.razor:
```diff
@@ -1,7 +1,8 @@
 @using Datahub.Core.Model.Projects
-@using Datahub.Core.Model.UserTracking
+@using Datahub.Core.Model.Users
+@using Datahub.Core.Model.Users
```
Duplicate using that Copilot missed

akhavich commented on Portal/test/Datahub.SpecflowTests/Steps/DatahubAzureSubscriptionSteps.cs:
```diff
@@ -69,7 +69,8 @@ public async Task WhenANewSubscriptionIsAdded()
         {
             SubscriptionId = Testing.WorkspaceSubscriptionGuid,
-            TenantId = Testing.WorkspaceTenantGuid
+            TenantId = Testing.WorkspaceTenantGuid,
```
This is repeated all over the place, maybe there should be a helper method or two. CreateTestingSubscription, CreateInvalidSubscription.

ErikApption replied:
agree

akhavich commented on Portal/test/Datahub.SpecflowTests/Steps/GCHosting/APICreateWorkspaceStepDefinitions.cs:
```diff
@@ -147,7 +148,7 @@ private static IWorkspaceCreationService CreateMockedWorkspaceCreationService(
 
             azureSubService.NextSubscriptionAsync()
-                .Returns(new Core.Model.Subscriptions.DatahubAzureSubscription() { Id = 1 });
+                .Returns(new Core.Model.Subscriptions.DatahubAzureSubscription() { Id = 1, SubscriptionId = "test-subscription-id", TenantId = "test-tenant-id", SubscriptionName = "Test Subscription" });
```
Another good place for CreateTestingSubscription

akhavich commented on Portal/test/Datahub.SpecflowTests/Steps/Workspace/CreateNewWorkspaceFormSteps.cs:
```diff
@@ -93,7 +93,7 @@ private static IWorkspaceCreationService CreateMockedWorkspaceCreationService(
 
             azureSubService.NextSubscriptionAsync()
-                .Returns(new Core.Model.Subscriptions.DatahubAzureSubscription() { Id = 1 });
+                .Returns(new Core.Model.Subscriptions.DatahubAzureSubscription() { Id = 1, SubscriptionId = "test-subscription-id", TenantId = "test-tenant-id", SubscriptionName = "Test Subscription" });
```
Another good place for CreateTestingSubscription

akhavich commented on Portal/test/Datahub.SpecflowTests/Utils/GenerateWorkspaceHelper.cs:
```diff
@@ -29,21 +38,56 @@ public static async Task GenerateWorkspace(IDbContextFactory<DatahubProjectDBCon
+            var rg = new Project_Resources2()
+            {
+                ResourceType = TerraformTemplate.GetTerraformServiceType(TerraformTemplate.NewProjectTemplate),
+                JsonContent = "{\"resource_group_name\":\"fsdh_proj_abc_dev_rg\"}",
```
Should use JSON_RG

ErikApption committed:
minor cleanup

akhavich reviewed (dismissed):
Looks good overall, some minor things could be improved

ErikApption committed:
removed duplicate line

ErikApption committed:
more fixing

ErikApption committed:
removed old test email

ErikApption committed:
removwed another duplicate

ErikApption committed:
fixed invalid email address

vb-ssc reviewed (dismissed):
Looks good to me

Sean-Stilwell reviewed (dismissed):
Working well locally. LGTM!

ErikApption committed:
- updated nullables for costing API
- added doc references
